### PR TITLE
Implementation of TIMER_FIRED, TIMER_SCHEDULED events

### PR DESCRIPTION
**Pull request by CTI777**
Related to https://github.com/Activiti/Activiti/issues/2636
Depends on https://github.com/Activiti/activiti-api/pull/121

### CODE REVIEW

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/ToJobExecutedSuccessConverter.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public Optional<BPMNTimerFiredEvent> from(ActivitiEntityEvent internalEvent) {
+        
+        JobEntityImpl jobEntity = (JobEntityImpl)internalEvent.getEntity();
+        
+        BPMNTimerImpl timer = new BPMNTimerImpl(jobEntity.getId());
+        timer.setProcessDefinitionId(internalEvent.getProcessDefinitionId());
+        timer.setProcessInstanceId(internalEvent.getProcessInstanceId());
+
+        TimerPayload timerPayload = ProcessPayloadBuilder.timer()
```
Comment (ffazzini) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/event/impl/ToJobExecutedSuccessConverter.java]: can this block `ProcessPayloadBuilder.timer()..` be moved to a common place for all converter classes?

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/event/internal/TimerCanceledListenerDelegate.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                                         ToTimerCanceledConverter converter) {
+        this.processRuntimeEventListeners = processRuntimeEventListeners;
+        this.converter = converter;
+    }
+
+    @Override
+    public void onEvent(ActivitiEvent event) {
+        if (event.getType().equals(ActivitiEventType.JOB_CANCELED)) {  
+        
+            if (event instanceof ActivitiEntityEvent &&
```
Comment (ffazzini) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/event/internal/TimerCanceledListenerDelegate.java]: can we wrap the multi condition if into a separate private method, choosing a method name which helps to understand what sort of event is filtering?

File: activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/event/internal/TimerExecutionSuccessListenerDelegate.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public void onEvent(ActivitiEvent event) {
+        if (event.getType().equals(ActivitiEventType.JOB_EXECUTION_SUCCESS)) {  
+        
+            if (event instanceof ActivitiEntityEvent && 
+                JobEntity.class.isAssignableFrom(((ActivitiEntityEvent) event).getEntity().getClass()) &&
+                ((JobEntityImpl)((ActivitiEntityEvent) event).getEntity()).getJobType().equals("timer")) {
+                converter.from((ActivitiEntityEvent) event)
+                        .ifPresent(convertedEvent -> {
+                            for (BPMNElementEventListener<BPMNTimerExecutionSuccessEvent> listener : processRuntimeEventListeners) {
```
Comment (ffazzini) [inline activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/event/internal/TimerExecutionSuccessListenerDelegate.java]: would be nice to have generic BPMNElementEventListener<T> so we could avoid this block of code repetition among different classes

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeBPMNTimerIT.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        
+    }
+
+    @After
+    public void tearDown() {
+        processCleanUpUtil.cleanUpWithAdmin();
+    }
+
+    @Test
+    public void shouldGetTimerFiredScheduledEventsForProcessWithTimer() {
```
Comment (ffazzini) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/process/ProcessRuntimeBPMNTimerIT.java]: maybe the first scenario can be splitted in two different scenarios, one without delay and one with delay?  That might help readability?

**Comment by erdemedeiros** [comment]
I'm closing this PR because it has some noisy of commits coming from other branches.
Comments that are already here should be handled at https://github.com/Activiti/Activiti/pull/2814

**Pull request closed without merging**